File: ontrack-extension-gitlab/src/main/java/net/nemerosa/ontrack/extension/gitlab/model/GitLabIssueWrapper.java
package net.nemerosa.ontrack.extension.gitlab.model;

import lombok.Data;
import net.nemerosa.ontrack.common.Time;
import net.nemerosa.ontrack.extension.issues.model.Issue;
import net.nemerosa.ontrack.extension.issues.model.IssueStatus;
import org.gitlab.api.models.GitlabIssue;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;

@Data
public class GitLabIssueWrapper implements Issue {

    private final GitlabIssue gitlabIssue;
    private final String milestoneUrl;
    private final String issueUrl;

    @Override
    public String getKey() {
        return String.valueOf(gitlabIssue.getId());
    }

    @Override
    public String getSummary() {
        return gitlabIssue.getTitle();
    }

    @Override
    public String getUrl() {
        return issueUrl;
    }

    @Override
    public IssueStatus getStatus() {
        return new GitLabIssueStatusWrapper(gitlabIssue.getState());
    }

    @Override
    public LocalDateTime getUpdateTime() {
        return Time.from(gitlabIssue.getUpdatedAt(), null);
    }

    public static GitLabIssueWrapper of(GitlabIssue issue, String milestoneUrl, String issueUrl) {
        return new GitLabIssueWrapper(issue, milestoneUrl, issueUrl);
    }

    public List<String> getLabels() {
        return Arrays.asList(gitlabIssue.getLabels());
    }
}

File: ontrack-extension-gitlab/src/main/java/net/nemerosa/ontrack/extension/gitlab/model/GitLabIssueWrapper.kt
package net.nemerosa.ontrack.extension.gitlab.model

import net.nemerosa.ontrack.common.Time
import net.nemerosa.ontrack.extension.issues.model.Issue
import net.nemerosa.ontrack.extension.issues.model.IssueStatus
import java.time.LocalDateTime

typealias GitLabIssue = org.gitlab4j.api.models.Issue

class GitLabIssueWrapper(
        val gitlabIssue: GitLabIssue,
        val milestoneUrl: String?
) : Issue {

    override val url: String = gitlabIssue.webUrl
    override val key: String = gitlabIssue.id.toString()
    override val displayKey: String = "#${gitlabIssue.id}"
    override val summary: String = gitlabIssue.title
    override val status: IssueStatus = GitLabIssueStatusWrapper(gitlabIssue.state.name)
    override val updateTime: LocalDateTime = Time.from(gitlabIssue.updatedAt.time)
    val labels: List<String> = gitlabIssue.labels

}